src/main/python/gemini/site_power_controller_utils.py

```diff
@@ -0,0 +1,470 @@
+# this file creates an intermediate federate
+# it maps the coordinates from TEMPO to the nearest charging
+# station modeled in PyDSS
+import time
+import helics as h
+import pandas as pd
+import logging
+import json
+import itertools
+import os
+import collections.abc
+
+from threading import Thread
+
+from rudimentary_spmc import SPM_Control
+
+
+class DefaultSPMC:
+    # Myungsoo is SPMC (NOT RIDE HAIL DEPOT)
+    def __init__(self, name, site_id):
+        self.site_id = site_id
+        self.site_prefix_logging = name + "[SITE:" + str(site_id) + "]. "
+        print2(self.site_prefix_logging + "Initializing the SPMCs...")
+        self.spm_c = SPM_Control(time_step_mins=1, max_power_evse=[], min_power_evse=[])
+
+    def log(self, log_message):
+        logging.info(self.site_prefix_logging + log_message)
+
+    def run(self, t, charging_events):
+        vehicle_id = []
+        vehicle_type = []
+        primary_fuel_level_in_k_wh = []
+        arrival_time = []
+        desired_departure_time = []
+        desired_fuel_level_in_k_wh = []
+        max_power_in_kw = []  # min [plug, vehicle]
+        battery_capacity_in_k_wh = []  # TODO Julius @ HL can you please add this to the BEAM output?
+        control_commands = []
+
+        for charging_event in charging_events:
+            vehicle_id.append(str(charging_event['vehicleId']))
+            vehicle_type.append(str(charging_event['vehicleType']))
+            # MJ: What is this? Is it the current energy level of each EV? Or battery size?
+            primary_fuel_level_in_k_wh.append(int(charging_event['primaryFuelLevelInJoules']) / 3600000)
+            # MJ: Should be in minutes of day. What is the unit of this?
+            arrival_time.append(int(charging_event['arrivalTime']))
+            # MJ: Should be in minutes of day. What is the unit of this?
+            desired_departure_time.append(int(charging_event['departureTime']))
+            # MJ: I assume that this is remaining energy to be delivered to each EV
+            # and updated each time,right?
+            desired_fuel_level_in_k_wh.append(int(charging_event['desiredFuelLevelInJoules']) / 3600000)
+            # MJ: I assume that this is the EV charging power
+            max_power_in_kw.append(float(charging_event['maxPowerInKW']))
+            # Julius @ HL can you please add this to the BEAM output?
+            battery_capacity_in_k_wh.append(int(charging_event['primaryFuelCapacityInJoule']) / 3600000)
```

Comment:
This event transforming code is presented 3 times in the project. Could we extract it?

Reply:
Unfortunately I don't see any simple way to refactor it, there is also lots of incomplete and commented-out code in the file, I bet it will be changed later.